### PR TITLE
[HttpClient] Allow enabling buffering conditionally with a Closure

### PR DESCRIPTION
**Pull request by rjwebdev**
| Q             | A
| ------------- | ---
| Branch?       | 4.4
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #31883 
| License       | MIT
| Doc PR        | symfony/symfony-docs#12043

With this PR, responses can be buffered automatically from a closure passed to the `buffer` option.

```php
$resp = $client->request('GET', $url, [
    'buffer' => function (array $headers): bool { return true/false; },
]);
```

When no option is provided, buffering is now enabled only for json, xml and text/* content types.

### CODE REVIEW

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php [hunk 1]
```diff
@@ -73,8 +73,8 @@ public function __construct(CurlClientState $multi, $ch, array $options = null,
             $content = ($options['buffer'] ?? true) ? $content : null;
         }
 
-        curl_setopt($ch, CURLOPT_HEADERFUNCTION, static function ($ch, string $data) use (&$info, &$headers, $options, $multi, $id, &$location, $resolveRedirect, $logger): int {
-            return self::parseHeaderLine($ch, $data, $info, $headers, $options, $multi, $id, $location, $resolveRedirect, $logger);
+        curl_setopt($ch, CURLOPT_HEADERFUNCTION, static function ($ch, string $data) use (&$info, &$headers, &$options, $multi, $id, &$location, $resolveRedirect, $logger, &$content): int {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/CurlResponse.php]: does $option really need to be passed by reference? why? better not otherwise

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php [hunk 2]
```diff
@@ -277,15 +277,19 @@ private static function select(CurlClientState $multi, float $timeout): int
...
     {
         if (!\in_array($waitFor = @curl_getinfo($ch, CURLINFO_PRIVATE), ['headers', 'destruct'], true)) {
             return \strlen($data); // Ignore HTTP trailers
         }
 
         if ("\r\n" !== $data) {
             // Regular header line: add it to the list
+
             self::addResponseHeaders([substr($data, 0, -2)], $info, $headers);
+            if (\array_key_exists('content-type', $headers)) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/CurlResponse.php]: isset() would be enough

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php [hunk 3]
```diff
@@ -235,6 +235,15 @@ private static function addResponseHeaders(array $responseHeaders, array &$info,
         }
     }
 
+    private static function handleContentType(&$content, string $contentType)
+    {
+        $contentType = substr($contentType, 0, strpos($contentType, ';'));
+        if (1 === preg_match('/(text\/[a-z]+|[a-z]+\/(json|xml)$)/', $contentType) && null === $content) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: instead of hardcoding the check, we should make it configurable:
when option "buffer" is set to a string, we could consider it a regexp that the content-type should match to enable buffering.
note that we don't use the `1 === ` part so it should be removed

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php [hunk 4]
```diff
@@ -235,6 +235,15 @@ private static function addResponseHeaders(array $responseHeaders, array &$info,
...
+            $options['buffer'] = true;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: undefined variable - but not need to set this here, this no need for passing it by ref (this was the intent I suppose)

**Review by nicolas-grekas — COMMENTED** [review]
Thanks for working on this :)

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php (hunk 1 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/CurlResponse.php]: Removed the by reference passing

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php (hunk 3 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: Provided a check to the buffer option if it's a string and removed the `1 ===` part

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php (hunk 4 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: Removed the unused variable

**Comment by rjwebdev** [comment]
@nicolas-grekas 

Could you please give me some tips to make sure the tests pass? I'm not sure how to fix them

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php [hunk 5]
```diff
@@ -277,15 +277,19 @@ private static function select(CurlClientState $multi, float $timeout): int
     /**
      * Parses header lines as curl yields them to us.
      */
-    private static function parseHeaderLine($ch, string $data, array &$info, array &$headers, ?array $options, CurlClientState $multi, int $id, ?string &$location, ?callable $resolveRedirect, ?LoggerInterface $logger): int
+    private static function parseHeaderLine($ch, string $data, array &$info, array &$headers, ?array &$options, CurlClientState $multi, int $id, ?string &$location, ?callable $resolveRedirect, ?LoggerInterface $logger, &$content = null): int
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/CurlResponse.php]: `&` should be removed for $options

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php [hunk 6]
```diff
@@ -277,15 +277,19 @@ private static function select(CurlClientState $multi, float $timeout): int
...
     {
         if (!\in_array($waitFor = @curl_getinfo($ch, CURLINFO_PRIVATE), ['headers', 'destruct'], true)) {
             return \strlen($data); // Ignore HTTP trailers
         }
 
         if ("\r\n" !== $data) {
             // Regular header line: add it to the list
+
             self::addResponseHeaders([substr($data, 0, -2)], $info, $headers);
+            if (isset($headers['content-type'], $options['buffer']) && \is_string($options['buffer'])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/CurlResponse.php]: this check should be done after, before `curl_setopt($ch, CURLOPT_PRIVATE, 'content');` below, so that we don't check the content-type for intermediary redirections

File: src/Symfony/Component/HttpClient/Response/NativeResponse.php [hunk 7]
```diff
@@ -130,6 +132,10 @@ private function open(): void
                 // Send request and follow redirects when needed
                 $this->handle = $h = fopen($url, 'r', false, $this->context);
                 self::addResponseHeaders($http_response_header, $this->info, $this->headers, $this->info['debug']);
+                if (null !== $this->bufferRegex && isset($this->headers['content-type'])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/NativeResponse.php]: same here: this should be outside the redirection loop

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php [hunk 8]
```diff
@@ -244,6 +244,14 @@ private static function addResponseHeaders(array $responseHeaders, array &$info,
         }
     }
 
+    private static function handleContentType(&$content, string $contentType, string $bufferRegex): void
+    {
+        $contentType = substr($contentType, 0, strpos($contentType, ';'));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: the semi-colon is optional: strpos can return false

File: src/Symfony/Component/HttpClient/Tests/CurlHttpClientTest.php [hunk 9]
```diff
@@ -71,4 +71,49 @@ public function log($level, $message, array $context = [])
         ];
         $this->assertSame($expected, $logger->logs);
     }
+
+    public function testJsonRegexBuffer(): void
+    {
+        $client = new CurlHttpClient(['buffer' => '/([a-z]+\/json$)/']);
+        $response = $client->request('GET', 'https://jsonplaceholder.typicode.com/users');
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Tests/CurlHttpClientTest.php]: instead of hitting external domains we don't control, we should hit the test server (see existing test cases)
also since both clients have the same behavior, we should be able to move the test cases to the parent class

**Review by nicolas-grekas — COMMENTED** [review]
Thanks :)
We also need to validate the regex and fail properly when it is not valid, can you add a test case ensuring so?

File: src/Symfony/Component/HttpClient/Tests/CurlHttpClientTest.php (hunk 9 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Tests/CurlHttpClientTest.php]: I moved the tests to the base unit test class and changed the remote server. Thanks for pointing to this local test server

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php (hunk 8 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: Added an extra check to see if strpos returns false. In that case, I'll use the length of the contentType variable

File: src/Symfony/Component/HttpClient/Response/NativeResponse.php (hunk 7 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/NativeResponse.php]: Moved it outside the loop

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php (hunk 6 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/CurlResponse.php]: Moved it

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php (hunk 5 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/CurlResponse.php]: Removed the by-reference pointer

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php (hunk 2 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/CurlResponse.php]: Changed to isset function

File: src/Symfony/Component/HttpClient/HttpClientTrait.php [hunk 10]
```diff
@@ -42,6 +42,14 @@ private static function prepareRequest(?string $method, ?string $url, array $opt
 
         $options = self::mergeDefaultOptions($options, $defaultOptions, $allowExtraOptions);
 
+        if (isset($options['buffer']) && \is_string($options['buffer'])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/HttpClientTrait.php]:
```php
if (\is_string($options['buffer'] ?? null) && false === @preg_match($options['buffer'], '')) {
    throw new InvalidArgumentException('Option "buffer" should be a valid regular expression.');
}
```

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php [hunk 11]
```diff
@@ -63,18 +63,18 @@ public function __construct(CurlClientState $multi, $ch, array $options = null,
         }
 
         if (null === $content = &$this->content) {
-            $content = ($options['buffer'] ?? true) ? fopen('php://temp', 'w+') : null;
+            $content = isset($options['buffer']) && true === $options['buffer'] ? fopen('php://temp', 'w+') : null;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/CurlResponse.php]: `($options['buffer'] ?? true) && !\is_string($options['buffer'] ?? true)` (same in other places)

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php [hunk 12]
```diff
@@ -277,14 +277,15 @@ private static function select(CurlClientState $multi, float $timeout): int
...
-    private static function parseHeaderLine($ch, string $data, array &$info, array &$headers, ?array $options, CurlClientState $multi, int $id, ?string &$location, ?callable $resolveRedirect, ?LoggerInterface $logger): int
+    private static function parseHeaderLine($ch, string $data, array &$info, array &$headers, ?array $options, CurlClientState $multi, int $id, ?string &$location, ?callable $resolveRedirect, ?LoggerInterface $logger, &$content = null): int
     {
         if (!\in_array($waitFor = @curl_getinfo($ch, CURLINFO_PRIVATE), ['headers', 'destruct'], true)) {
             return \strlen($data); // Ignore HTTP trailers
         }
 
         if ("\r\n" !== $data) {
             // Regular header line: add it to the list
+
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/CurlResponse.php]: unrelated, should be removed

File: src/Symfony/Component/HttpClient/Response/NativeResponse.php [hunk 13]
```diff
@@ -50,7 +51,8 @@ public function __construct(NativeClientState $multi, $context, string $url, $op
         $this->info = &$info;
         $this->resolveRedirect = $resolveRedirect;
         $this->onProgress = $onProgress;
-        $this->content = $options['buffer'] ? fopen('php://temp', 'w+') : null;
+        $this->content = isset($options['buffer']) && true === $options['buffer'] ? fopen('php://temp', 'w+') : null;
+        $this->bufferRegex = isset($options['buffer']) && \is_string($options['buffer']) ? $options['buffer'] : null;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/NativeResponse.php]: `\is_string($options['buffer'] ?? null)`

File: src/Symfony/Component/HttpClient/Response/NativeResponse.php [hunk 14]
```diff
@@ -130,6 +132,7 @@ private function open(): void
                 // Send request and follow redirects when needed
                 $this->handle = $h = fopen($url, 'r', false, $this->context);
                 self::addResponseHeaders($http_response_header, $this->info, $this->headers, $this->info['debug']);
+
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/NativeResponse.php]: unrelated

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php [hunk 15]
```diff
@@ -244,6 +244,14 @@ private static function addResponseHeaders(array $responseHeaders, array &$info,
         }
     }
 
+    private static function handleContentType(&$content, string $contentType, string $bufferRegex): void
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: let's not pass $content at all and name the method createBufferForContentType?
this means checking for null === $content outside of the method of course

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php [hunk 16]
```diff
@@ -244,6 +244,14 @@ private static function addResponseHeaders(array $responseHeaders, array &$info,
...
+    {
+        $contentType = substr($contentType, 0, (false !== strpos($contentType, ';') ? strpos($contentType, ';') : \strlen($contentType)));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/ResponseTrait.php]:
```php
if (false !== $i = strpos($contentType, ';')) {
    $contentType = substr($contentType, 0, $i);
}

return preg_match($bufferRegex, strtolower($contentType)) ? fopen('php://temp', 'w+') : null;
```

File: src/Symfony/Component/HttpClient/Tests/MockHttpClientTest.php [hunk 17]
```diff
@@ -122,6 +122,9 @@ protected function getHttpClient(string $testCase): HttpClientInterface
             case 'testUncheckedTimeoutThrows':
                 $body = ['<1>', '', '<2>'];
                 $responses[] = new MockResponse($body, ['response_headers' => $headers]);
+                break;
+            case 'testBufferRegex':
+
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Tests/MockHttpClientTest.php]: should we mark the test as skipped? (as others are above)

File: src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php [hunk 18]
```diff
@@ -763,4 +763,62 @@ public function testGzipBroken()
...
+        $firstContent = $response->getContent();
+        $secondContent = $response->getContent();
+
+        $this->assertSame($firstContent, $secondContent);
+    }
+
+    /**
+     * @expectedException \Symfony\Component\HttpClient\Exception\TransportException
+     */
+    public function testFalseBuffer(): void
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php]: this is already tested in `testNonBufferedGetRequest`

File: src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php [hunk 19]
```diff
@@ -763,4 +763,62 @@ public function testGzipBroken()
...
+    public function testPdfRegexBuffer(): void
+    {
+        $client = $this->getHttpClient(__FUNCTION__);
+        $response = $client->request('GET', 'http://localhost:8057', ['buffer' => '/([a-z]+\/json))/']);
+
+        $firstContent = $response->getContent();
+        $secondContent = $response->getContent();
+    }
+
+    public function testTrueBuffer(): void
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php]: this is already tested, although indirectly

File: src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php [hunk 20]
```diff
@@ -763,4 +763,62 @@ public function testGzipBroken()
         $this->expectException(TransportExceptionInterface::class);
         $response->getContent();
     }
+
+    public function testJsonRegexBuffer(): void
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php]: this should be moved to `HttpClientTestCase` in the `Symfony\Component\HttpClient\Tests` namespace
(as the new behavior is not in the contracts)

File: src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php [hunk 21]
```diff
@@ -763,4 +763,62 @@ public function testGzipBroken()
...
+        $response = $client->request('GET', 'http://localhost:8057', ['buffer' => false]);
+
+        $firstPosts = $response->getContent();
+        $secondPosts = $response->getContent();
+    }
+
+    /**
+     * @expectedException \Symfony\Component\HttpClient\Exception\InvalidArgumentException
+     */
+    public function testInvalidRegexBuffer(): void
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php]: should be moved too

File: src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php [hunk 22]
```diff
@@ -763,4 +763,62 @@ public function testGzipBroken()
...
+        $response = $client->request('GET', 'http://localhost:8057', ['buffer' => '/([a-z]+\/json))/']);
+
+        $firstPosts = $response->getContent();
+        $secondPosts = $response->getContent();
+    }
+
+    /**
+     * @expectedException \Symfony\Component\HttpClient\Exception\InvalidArgumentException
+     */
+    public function testPdfRegexBuffer(): void
```
Comment (nicolas-grekas) [inline src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php]: should be moved too (but I don't get what this tests?)

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php (hunk 15 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: You mean that I should return the stream created by `fopen` in the createBufferForContentType or otherwise `null` if regex matches?

File: src/Symfony/Component/HttpClient/HttpClientTrait.php (hunk 10 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/HttpClientTrait.php]: Replaced the check and changed the exception message

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php (hunk 11 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/CurlResponse.php]: Changed the check on all places

File: src/Symfony/Component/HttpClient/Response/NativeResponse.php (hunk 14 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/NativeResponse.php]: Removed empty line

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php (hunk 15 above)
Comment (nicolas-grekas) [reply src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: Correct

File: src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php (hunk 18 above)
Comment (rjwebdev) [reply src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php]: Removed the test

File: src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php (hunk 20 above)
Comment (rjwebdev) [reply src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php]: Moved the test to the correct class

File: src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php (hunk 22 above)
Comment (rjwebdev) [reply src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php]: Removed the test since it didn't have a meaning

File: src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php (hunk 21 above)
Comment (rjwebdev) [reply src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php]: Moved to the correct test class

File: src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php (hunk 19 above)
Comment (rjwebdev) [reply src/Symfony/Contracts/HttpClient/Test/HttpClientTestCase.php]: Removed this test

File: src/Symfony/Component/HttpClient/Response/ResponseTrait.php (hunk 15 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/ResponseTrait.php]: Since there's no return type for a stream (yet?), can/should I add a phpdoc block to declare the return type?

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php [hunk 23]
```diff
@@ -63,18 +63,18 @@ public function __construct(CurlClientState $multi, $ch, array $options = null,
...
-            $content = ($options['buffer'] ?? true) ? fopen('php://temp', 'w+') : null;
+            $content = ($options['buffer'] ?? true) && !\is_string($options['buffer'] ?? true) ? fopen('php://temp', 'w+') : null;
         } else {
             // Move the pushed response to the activity list
             if (ftell($content)) {
                 rewind($content);
                 $multi->handlesActivity[$id][] = stream_get_contents($content);
             }
-            $content = ($options['buffer'] ?? true) ? $content : null;
+            $content = isset($options['buffer']) && true === $options['buffer'] ? $content : null;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/CurlResponse.php]: `($options['buffer'] ?? true) && !\is_string($options['buffer'] ?? true)` too

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php [hunk 24]
```diff
@@ -345,6 +345,10 @@ private static function parseHeaderLine($ch, string $data, array &$info, array &
                 return 0;
             }
 
+            if (isset($headers['content-type'], $options['buffer']) && \is_string($options['buffer'])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/CurlResponse.php]: `if (!$content && [...]`

File: src/Symfony/Component/HttpClient/Response/NativeResponse.php [hunk 25]
```diff
@@ -152,6 +154,10 @@ private function open(): void
         stream_set_blocking($h, false);
         $this->context = $this->resolveRedirect = null;
 
+        if (null !== $this->bufferRegex && isset($this->headers['content-type'])) {
```
Comment (nicolas-grekas) [inline src/Symfony/Component/HttpClient/Response/NativeResponse.php]: `if (null !== $this->content && [...]`

**Review by nicolas-grekas — COMMENTED** [review]
(no need for the return-type annotation on createBufferForContentType IMHO)

File: src/Symfony/Component/HttpClient/Response/NativeResponse.php (hunk 25 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/NativeResponse.php]: Fixed this already

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php (hunk 23 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/CurlResponse.php]: Changed

File: src/Symfony/Component/HttpClient/Response/CurlResponse.php (hunk 24 above)
Comment (rjwebdev) [reply src/Symfony/Component/HttpClient/Response/CurlResponse.php]: Changed the check. Should I change the order of the conditions too?

**Comment by nicolas-grekas** [comment]
Thanks, LGTM for the component.
Note that the PR should target 4.4, can you rebase+retarget?
If not, don't worry we can do it while merging.

Could you please have a look at the configuration for the component in FrameworkBundle?
I think the option "buffer" is defined as boolean there, so we should relax this now, and add a regexp check to ensure it is valid at compile time.

**Comment by rjwebdev** [comment]
Right now the default value of the buffer option is true, defined in `Symfony\Contracts\HttpClient\HttpClientInterface`. There's no buffer option configured in the frameworkbundle yet. 

Should I add this option in the configuration (both default_options and scoped_clients?) and validate it in the configuration?

**Comment by nicolas-grekas** [comment]
I'd say yes :)

**Comment by rjwebdev** [comment]
I added the buffer configuration option for both the default_options as the scoped_clients.

Also updated the documentation in a PR

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 26]
```diff
@@ -1393,6 +1393,13 @@ private function addHttpClientSection(ArrayNodeDefinition $rootNode)
                                 ->scalarNode('ciphers')
                                     ->info('A list of SSL/TLS ciphers separated by colons, commas or spaces (e.g. "RC3-SHA:TLS13-AES-128-GCM-SHA256"...)')
                                 ->end()
+                                ->scalarNode('buffer')
+                                    ->info('Response content should be buffered or not')
+                                    ->validate()
+                                        ->ifTrue(function ($v) { return !\is_bool($v) && (\is_string($v) && false === @preg_match($v, '')); })
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: `!\is_bool($v) && (!\is_string($v) || false === @preg_match($v, ''))`

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 27]
```diff
@@ -1393,6 +1393,13 @@ private function addHttpClientSection(ArrayNodeDefinition $rootNode)
                                 ->scalarNode('ciphers')
                                     ->info('A list of SSL/TLS ciphers separated by colons, commas or spaces (e.g. "RC3-SHA:TLS13-AES-128-GCM-SHA256"...)')
                                 ->end()
+                                ->scalarNode('buffer')
+                                    ->info('Response content should be buffered or not')
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Whether the response content should be buffered or not or a regex telling which content-types should be

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 28]
```diff
@@ -503,6 +503,11 @@
         <xsd:attribute name="local-pk" type="xsd:string" />
         <xsd:attribute name="passphrase" type="xsd:string" />
         <xsd:attribute name="ciphers" type="xsd:string" />
+        <xsd:attribute name="buffer">
+            <xsd:simpleType>
+                <xsd:union memberTypes="xsd:string xsd:boolean" />
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: I think `type="xsd:string"` would cover all cases

Comment (rjwebdev) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: Simplified the configuration. Looks like it's working

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 27 above)
Comment (rjwebdev) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Changed the info content

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 26 above)
Comment (rjwebdev) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Changed the check

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 29]
```diff
@@ -1393,6 +1393,13 @@ private function addHttpClientSection(ArrayNodeDefinition $rootNode)
                                 ->scalarNode('ciphers')
                                     ->info('A list of SSL/TLS ciphers separated by colons, commas or spaces (e.g. "RC3-SHA:TLS13-AES-128-GCM-SHA256"...)')
                                 ->end()
+                                ->scalarNode('buffer')
+                                    ->info('Whether the response content should be buffered or a regex telling which content-types should be')
+                                    ->validate()
+                                        ->ifTrue(function ($v) { return !\is_bool($v) && (!\is_string($v) || false === @preg_match($v, '')); })
+                                        ->thenInvalid('Buffer option should be a boolean or a valid regex')
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: please add a final dot

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php [hunk 30]
```diff
@@ -1529,6 +1536,13 @@ private function addHttpClientSection(ArrayNodeDefinition $rootNode)
                                     ->scalarNode('ciphers')
                                         ->info('A list of SSL/TLS ciphers separated by colons, commas or spaces (e.g. "RC3-SHA:TLS13-AES-128-GCM-SHA256"...)')
                                     ->end()
+                                    ->scalarNode('buffer')
+                                        ->info('Whether the response content should be buffered or a regex telling which content-types should be')
+                                        ->validate()
+                                            ->ifTrue(function ($v) { return !\is_bool($v) && (!\is_string($v) || false === @preg_match($v, '')); })
+                                            ->thenInvalid('Buffer option should be a boolean or a valid regex')
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: final dot

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 31]
```diff
@@ -532,6 +533,11 @@
...
```
Comment (nicolas-grekas) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: should be same as above

**Review by nicolas-grekas — COMMENTED** [review]
last comments on my side :)
@joelwurtz could you please have a look? WDYT?

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd (hunk 31 above)
Comment (rjwebdev) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: Changed type

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 30 above)
Comment (rjwebdev) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: added

File: src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php (hunk 29 above)
Comment (rjwebdev) [reply src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php]: Final dot added

**Comment by joelwurtz** [comment]
Regex is nice to cover a lot of use case, I'm just wondering about the DX if this is the best option ? 

If i want to create the regex that cover all of the text types it's not an easy one to create but this may be resolve by providing good examples in the documentation ?

**Comment by nicolas-grekas** [comment]
can you please rebase to get rid of the conflict? we removed the ForwardCompatTestTrait recently.

**Comment by rjwebdev** [comment]
Should be ok now

**Review by nicolas-grekas — APPROVED** [review]

**Comment by fabpot** [comment]
I would not make the regex configurable (as I don't see the use cases but I can see how one can mess up with changing the regex). The `buffer` option should be `true`, `false`,  `auto` IMHO.

**Comment by rjwebdev** [comment]
Ok. So what should happen with this PR?

**Comment by fabpot** [comment]
We are on the right track. The configuration option should only allow `true`, `false`, `auto` (which is going to use you regex).

**Comment by rjwebdev** [comment]
Ok. Any idea which content types we'll allow to be buffered automatically? I need this one to be able to build a regex

File: src/Symfony/Component/HttpClient/CHANGELOG.md [hunk 32]
```diff
@@ -10,6 +10,7 @@ CHANGELOG
  * added `$response->toStream()` to cast responses to regular PHP streams
  * made `Psr18Client` implement relevant PSR-17 factories and have streaming responses
  * added `max_duration` option
+ * added support for regexp for the buffer option
```
Comment (OskarStark) [inline src/Symfony/Component/HttpClient/CHANGELOG.md]:
```suggestion
 * added support for regexp for the `buffer` option
```

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd [hunk 33]
```diff
@@ -534,6 +535,7 @@
         <xsd:attribute name="local-pk" type="xsd:string" />
         <xsd:attribute name="passphrase" type="xsd:string" />
         <xsd:attribute name="ciphers" type="xsd:string" />
+        <xsd:attribute name="buffer" type="xsd:string" />
```
Comment (OskarStark) [inline src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: is string correct here or is it because of the true/false **or** regex option?

**Comment by OskarStark** [comment]
Why not

```yaml
buffer: true
buffer_regex: ....
```

buffer regex is containing the default one which could be overidden and is used if buffer is `true`... 🤔 The type mixing (`bool`/`string`) is not so the best option IMHO

File: src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd (hunk 33 above)
Comment (rjwebdev) [reply src/Symfony/Bundle/FrameworkBundle/Resources/config/schema/symfony-1.0.xsd]: String is correct. When you pass true or false, it is handled correctly

**Review by nicolas-grekas — APPROVED** [review]
I updated this PR to take @fabpot's comment into account. The option "buffer" is now removed from FrameworkBundle.
Implementations in the component (not the contracts) support a `string` for the "buffer" option.
This string should be a regular expression that enables buffering when the content-type of the response matches it.
It defaults to `#^(?:text/|application/(?:.+\+)?(?:json|xml)$)#i`, which covers all common content-types that should be buffered to me.

**Comment by nicolas-grekas** [comment]
I figured out we could make this much more powerful by accepting a `Closure` to enable/disable buffering based on response headers. So there we are, PR updated.

@rjwebdev could you please sync the PR descrption and consider a documentation PR?

**Comment by rjwebdev** [comment]
@nicolas-grekas 

Since you removed the `buffer` option from the Framework bundle, shouldn't the PR for the docs be closed?

**Comment by nicolas-grekas** [comment]
We'd need a new PR, to explain the option, in the doc for the component: `components/http_client.rst`

**Comment by nicolas-grekas** [comment]
Thank you @rjwebdev.
